vendor/bat-native-ads/src/bat/ads/internal/ml/pipeline/pipeline_util.cc

```diff
@@ -0,0 +1,220 @@
+    return base::nullopt;
+  }
+
+  base::Optional<TransformationVector> transformations = TransformationVector();
+  auto transformation_list = transformations_value->GetList();
+  for (size_t i = 0; i < transformation_list.size(); i++) {
+    const base::Value& transformation = transformation_list[i];
+
+    const std::string* transformation_type =
+        transformation.FindStringKey("transformation_type");
+
+    if (!transformation_type) {
+      return base::nullopt;
+    }
+
+    std::string parsed_transformation_type = *transformation_type;
+
+    if (parsed_transformation_type.compare("TO_LOWER") == 0) {
+      transformations.value().push_back(
+          std::make_unique<LowercaseTransformation>(LowercaseTransformation()));
+    }
+
+    if (parsed_transformation_type.compare("NORMALIZE") == 0) {
+      transformations.value().push_back(
+          std::make_unique<NormalizationTransformation>(
+              NormalizationTransformation()));
+    }
+
+    if (parsed_transformation_type.compare("HASHED_NGRAMS") == 0) {
+      const base::Value* transformation_params =
+          transformation.FindKey("params");
+
+      if (!transformation_params) {
+        return base::nullopt;
+      }
+
+      const base::Optional<int> nb =
+          transformation_params->FindIntKey("num_buckets");
+      if (!nb.has_value()) {
+        return base::nullopt;
+      }
+
+      int num_buckets = nb.value();
+
+      const base::Value* n_gram_sizes =
+          transformation_params->FindListKey("ngrams_range");
+
+      if (!n_gram_sizes) {
+        return base::nullopt;
+      }
+
+      std::vector<int> ngram_range;
+      auto ngram_range_list = n_gram_sizes->GetList();
+      for (size_t i = 0; i < ngram_range_list.size(); i++) {
+        const base::Value& n = ngram_range_list[i];
+        ngram_range.push_back(n.GetInt());
+      }
+      HashedNGramsTransformation hashed_ngrams(num_buckets, ngram_range);
+      transformations.value().push_back(
+          std::make_unique<HashedNGramsTransformation>(hashed_ngrams));
+    }
+  }
+
+  return transformations;
+}
+
+base::Optional<model::Linear> ParseClassifierJSON(
+    base::Value* classifier_value) {
+  if (!classifier_value) {
+    return base::nullopt;
+  }
+
+  std::vector<std::string> classes;
+  std::string* classifier_type =
+      classifier_value->FindStringKey("classifier_type");
+
+  if (!classifier_type) {
+    return base::nullopt;
+  }
+
+  std::string parsed_classifier_type = *classifier_type;
+
+  if (parsed_classifier_type.compare("LINEAR")) {
+    return base::nullopt;
+  }
+
+  base::Value* specified_classes = classifier_value->FindListKey("classes");
+  if (!specified_classes) {
+    return base::nullopt;
+  }
+
+  auto specified_classes_list = specified_classes->GetList();
+  for (size_t i = 0; i < specified_classes_list.size(); i++) {
+    const base::Value& class_name = specified_classes_list[i];
+    classes.push_back(class_name.GetString());
```

Comment:
We should always check the type of object in base::Value, i.e. DCHECK if should be picked up during development or conditional statement if with NOTREACHED if could happen in the wild

Reply:
Done. Now return optional without value to denote that the pipeline is not initialized.